### DISCUSSION TITLE
bfloat16

### DISCUSSION THREAD

**Discussion by jeyblu** [onnx/onnx#4139] [Q&A]
Is it possible for onnx to support bfloat16 datatype? Thanks

Line [16](https://github.com/onnx/onnx/blob/main/onnx/mapping.py#L16) of onnx/onnx/mapping.py and lines [331-333](https://github.com/onnx/onnx/blob/main/onnx/helper.py#L332) of onnx/onnx/helper.py show that bfloat16 is recast to float16.

`int(TensorProto.BFLOAT16): np.dtype('float16'),  # native numpy does not support bfloat16`

```
elif (data_type == TensorProto.FLOAT16
        or data_type == TensorProto.BFLOAT16):
    vals = np.array(vals).astype(np.float16).view(dtype=np.uint16).flatten().tolist()
```

**Comment by jcwchen** [comment]
Hi @jeyblu,
ONNX does support bfloat16 as datatype and there will be more and more operators support bfloat16 (https://github.com/onnx/onnx/issues/3842) so I don't quite follow here -- what kind of support are you looking for? Are you saying assigning NumPy bfloat16 values support? IIUC, native NumPy has not supported bfloat16 yet.

**Comment by jeyblu** [comment]
Hi @jcwchen, yes assigning numpy bfloat16 values. We want to be able to use onnx/onnx/mapping.py and onnx/onnx/helper.py to generate bfloat16 models, e.g. bfloat16 weights. If numpy doesn't currently support bfloat16, how do you create onnx bfloat16 models? Thanks

**Reply by jcwchen** [reply]
Bfloat16 ONNX models come from TensorFlow so I think typically people will create such a model in TensorFlow with data type bfloat16 and then use tf2onnx to convert it to ONNX.

**Comment by jeyblu** [comment]
How do you convert fp32 onnx models to bfloat16 onnx models?

**Reply by jcwchen** [reply]
A recent PR has brought this support: https://github.com/onnx/onnx/pull/4193. You can check this function: [float32_to_bfloat16()](https://github.com/onnx/onnx/blob/cd23ca602c1b319b16020a2df453cd49011a0b6b/onnx/helper.py#L286).

**Comment by jeyblu** [comment]
Is bfloat16 stored as uint16 as suggested in line [16](https://github.com/onnx/onnx/blob/main/onnx/mapping.py#L16) of mapping.py?

For float16, lines [354-355](https://github.com/onnx/onnx/blob/cd23ca602c1b319b16020a2df453cd49011a0b6b/onnx/helper.py#L354) of helper.py uses astype(np.float16).view(dtype=np.uint16). But for bfloat16, lines [356-357](https://github.com/onnx/onnx/blob/cd23ca602c1b319b16020a2df453cd49011a0b6b/onnx/helper.py#L356) of helper.py uses  astype(np.float32). Is this referring to the source float32 tensor from which the data is converted or the target bfloat16 tensor?

**Reply by jcwchen** [reply]
> Is bfloat16 stored as uint16 as suggested in line [16](https://github.com/onnx/onnx/blob/main/onnx/mapping.py#L16) of mapping.py?

Yes. Thank you for questioning this so I might find an issue in the map: https://github.com/onnx/onnx/pull/4234.

> For float16, lines [354-355](https://github.com/onnx/onnx/blob/cd23ca602c1b319b16020a2df453cd49011a0b6b/onnx/helper.py#L354) of helper.py uses astype(np.float16).view(dtype=np.uint16). But for bfloat16, lines [356-357](https://github.com/onnx/onnx/blob/cd23ca602c1b319b16020a2df453cd49011a0b6b/onnx/helper.py#L356) of helper.py uses astype(np.float32). Is this referring to the source float32 tensor from which the data is converted or the target bfloat16 tensor?

Yes, make_tensor will take numpy.float32 as vals for TensorProto.bfloat16. Then it will convert given float32 to uint16 and store uint16 eventually in the tensor.

**Comment by jeyblu** [comment]
If numpy supports bfloat16 eventually, then this workaround should be changed, right? May be helpful to add some documentation to help and remind whoever needs to revert this workaround later.

**Reply by jcwchen** [reply]
Good point. I will add a TODO in the code to reminder developers in the PR: https://github.com/onnx/onnx/pull/4270 and file an issue in ONNX for future tracking: https://github.com/onnx/onnx/issues/4273.

**Reply by jcwchen** [reply]
To clarify, probably ONNX will keep both ways (np.bfloat16 and np.float32) for compatibility right after NumPy has supported np.bfloat16 and deprecate np.float32 for making bfloat16 tensors in the later future.
